lib/AST/AutoDiff.cpp

```diff
@@ -34,6 +34,95 @@ AutoDiffAssociatedFunctionKind(StringRef string) {
   rawValue = *result;
 }
 
+// TODO(TF-874): This helper is inefficient and should be removed. Unwrapping at
+// most once (for curried method types) is sufficient.
+static void unwrapCurryLevels(AnyFunctionType *fnTy,
+                              SmallVectorImpl<AnyFunctionType *> &result) {
+  while (fnTy != nullptr) {
+    result.push_back(fnTy);
+    fnTy = fnTy->getResult()->getAs<AnyFunctionType>();
+  }
+}
+
+static unsigned countNumFlattenedElementTypes(Type type) {
+  if (auto *tupleTy = type->getCanonicalType()->getAs<TupleType>())
+    return accumulate(tupleTy->getElementTypes(), 0,
+                      [&](unsigned num, Type type) {
+                        return num + countNumFlattenedElementTypes(type);
+                      });
+  return 1;
+}
+
+// TODO(TF-874): Simplify this helper. See TF-874 for WIP.
+AutoDiffIndexSubset *
+autodiff::getLoweredParameterIndices(AutoDiffIndexSubset *indices,
+                                     AnyFunctionType *type) {
+  SmallVector<AnyFunctionType *, 2> curryLevels;
+  unwrapCurryLevels(type, curryLevels);
+
+  // Calculate the lowered sizes of all the parameters.
+  SmallVector<unsigned, 8> paramLoweredSizes;
+  unsigned totalLoweredSize = 0;
+  auto addLoweredParamInfo = [&](Type type) {
+    unsigned paramLoweredSize = countNumFlattenedElementTypes(type);
+    paramLoweredSizes.push_back(paramLoweredSize);
+    totalLoweredSize += paramLoweredSize;
+  };
+  for (auto *curryLevel : reversed(curryLevels))
+    for (auto &param : curryLevel->getParams())
+      addLoweredParamInfo(param.getPlainType());
+
+  // Construct the result by setting each range of bits that corresponds to each
+  // "on" parameter.
+  llvm::SmallVector<unsigned, 8> loweredIndices;
+  unsigned currentBitIndex = 0;
+  for (unsigned i : range(indices->getCapacity())) {
+    auto paramLoweredSize = paramLoweredSizes[i];
+    if (indices->contains(i)) {
+      auto indices = range(currentBitIndex, currentBitIndex + paramLoweredSize);
+      loweredIndices.append(indices.begin(), indices.end());
+    }
+    currentBitIndex += paramLoweredSize;
+  }
+
+  return AutoDiffIndexSubset::get(
+      type->getASTContext(), totalLoweredSize, loweredIndices);
+}
+
+// TODO(TF-874): Simplify this helper and remove the `reverseCurryLevels` flag.
+// See TF-874 for WIP.
+void autodiff::getSubsetParameterTypes(AutoDiffIndexSubset *indexSet,
```

Comment:
```suggestion
void autodiff::getSubsetParameterTypes(AutoDiffIndexSubset *subset,
```

Reply:
Done in https://github.com/apple/swift/pull/27514/commits/37d1a911cd33f0a3bc45ce1fe0499727ed04c60a.